Add theme compatibility for Lesson page

alexsanford opened the pull request:
Partial solution to #2154

Adds theme compatibility for the Lesson page.

This PR extracts a siginificant amount of functionality from https://github.com/Automattic/sensei/pull/2155 and reuses it for Lessons. Please double check the Course page and Shortcode to ensure nothing got broken in the process.

## Testing

To test, simply load a Lesson page on several different themes. Make sure to try themes that support Sensei, and themes that do not. Also try with themes that Sensei provides integration for. Also, please double-check Course pages, and pages using the `sensei_course_page` shortcode.

donnapep commented:
## Divi Theme
- I've added a featured image to a course and a lesson. On the course and the lesson page, that featured image shows twice.
- For a particular lesson, select _Discussion_ > _Allow comments_ checkbox. (Check _Discussion_ in _Screen Options_ first if not visible). Navigate to that lesson on the front-end. The comments box shows twice.

## Jupiter Theme
- The layout for the lesson page is still broken.
- Duplicate title and featured image on course page.
- When you're ready I can give you a copy of this theme.

## All Themes
- Activate the Sensei Media Attachments plugin and attach course media to a course. Navigate to the course page. The course media now appears before the course description, whereas before it appeared after the list of lessons.

jom commented:
I was able to reproduce the featured image issue @donnapep found on lesson or course pages with Divi 3.10.1 only if the `Output the [Course|Lesson] Image on the Single [Course|Lesson] Page` Sensei setting(s) are enabled. It might just be recommending them to disable that setting when their theme supports featured images.
Without setting enabled:
<img width="1983" alt="screen shot 2018-09-05 at 11 46 10 am" src="https://user-images.githubusercontent.com/68693/45088672-a5d3b980-b101-11e8-80e0-03554ecbca3f.png">

With setting enabled:
<img width="1983" alt="screen shot 2018-09-05 at 11 52 30 am" src="https://user-images.githubusercontent.com/68693/45088859-3ca07600-b102-11e8-9445-fefdf76d2c7d.png">

For the discussion bug, I don't see that in my Screen Options:
<img width="1983" alt="screen shot 2018-09-05 at 11 59 04 am" src="https://user-images.githubusercontent.com/68693/45089153-34950600-b103-11e8-9f6f-2fcf03b4e159.png">

alexsanford commented:
FYI rebased this and changed the merge base to `master` since https://github.com/Automattic/sensei/pull/2205 is merged.

alexsanford commented:
@donnapep @jom I've pushed some new commits here to address the issues that y'all brought up.

**Divi Issues:**

> I've added a featured image to a course and a lesson. On the course and the lesson page, that featured image shows twice regardless of whether or not you are logged in.

As discussed in our team chat, the solution for now will be to instruct users on how to disable the featured image display within Sensei. This way, if the theme displays featured images, they can hide Sensei's duplicate.

> For a particular lesson, select Discussion > Allow comments checkbox. (Check Discussion in Screen Options first if not visible). Navigate to that lesson on the front-end. The comments box shows twice when logged in.

This is fixed in 8d11b9c3569803fafd56ee5833a1fd199723c285.

**Jupiter Issues:**

> The layout for the lesson page is still broken.

I believe that bf32e13937e77b8e010aff5f34873f8a434a4544 will fix this. Please update to the latest on this branch and let me know if it's fixed on your end.

> Duplicate title and featured image on course page.

See above for featured image issue.

For the title, Jupiter has settings on each post called "Jupiter Page Options". One of these options is "Manage Page Elements" and in the dropdown you can choose to "Remove Page Title". This fixes the issue.

Also note that this same issue (and solution) exists for regular Posts. Consequently, I'm going to say that this is _not_ a theme compatibility issue.

> Activate the Sensei Media Attachments plugin and attach course media to a course. Navigate to the course page. The course media now appears before the course description, whereas before it appeared after the list of lessons.

The media links should never appear below the list of lessons. I cannot make it do so in any of my test environments. Furthermore, it happens on @donnapep's sandbox even on a supported theme (Twenty Sixteen). I do not believe this is a problem with the theme compatibility code; I'm guessing there's something weird going on with that setup. But if someone can provide compelling evidence otherwise, with some idea of how to reproduce it, I can dig in deeper.

(@donnapep to confirm, maybe you could try with Sensei 1.11.0 on your sandbox vs. a local version and see whether the issue exists there?)

**Code Review:**

@jom I've addressed your code comments.

As discussed in the team chat today, we'll hold off on creating a setting for pagination in favour of sticking with the filter for now.

---

Alright, this should be ready for testing again :slightly_smiling_face:

jom commented:
@alexsanford Looks good!

donnapep commented:
@alexsanford Don't forget to delete your branches after merging. 🙂

alexsanford commented:
Whoops. Thanks for the reminder! 🙂
